File: Codebase/Trinity Link MK3/Trinity/Classes/Comms/Comms.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using System.IO;
using System.IO.Ports;
using System.Threading;
using System.Windows.Forms;

namespace Trinity
{

    public class Comms
    {
        const byte TRANSMISSION_IN_NONE = 0;
        const byte TRANSMISSION_IN_LEDMANAGER = 1;
        const byte TRANSMISSION_IN_PANEL = 2;
        const byte TRANSMISSION_IN_DIODE = 3;
        const byte TRANSMISSION_IN_IDENT = 4;

        const byte TRANSMISSION_OUT_NONE         = 0;
        const byte TRANSMISSION_OUT_PANELFX      = 1;
        const byte TRANSMISSION_OUT_PANELCUSTOM  = 2;
        const byte TRANSMISSION_OUT_DIODEFX      = 3;
        const byte TRANSMISSION_OUT_DIODECUSTOM  = 4;
        const byte TRANSMISSION_OUT_BRIGHTNESS   = 5;
        const byte TRANSMISSION_OUT_SLEEPTIMER   = 6;
        const byte TRANSMISSION_OUT_LIGHTSENSOR  = 7;
        const byte TRANSMISSION_OUT_REQUEST      = 8;
        const byte TRANSMISSION_OUT_IDENT        = 9;

        const byte AUTOCONNECT_FAILED   = 0;
        const byte AUTOCONNECT_SUCCESS  = 1;
        const byte AUTOCONNECT_BUSY     = 2;
        const byte BUFFERSIZE = 5;
        const byte TIMEOUTATTEMPTS = 20;

        bool connected;
        byte readyTransmissionType;
        byte timeoutTimer;
        byte[] transmissionData = new byte[BUFFERSIZE];
        SerialPort serial = new SerialPort();
        Transmission_LedManager buffer_ledManager;
        Transmission_Panel buffer_panel;
        Transmission_Diode buffer_diode;
        string[] ports;
        int currentConnectingPort = -1; //-1 Because a ++ will set it to 0
        bool currentlyConnecting;

        //Constructor
        public Comms()
        {
        }
        //Private
        public byte decodeTransmissionType()
        {
            if (
                transmissionData[0] == 'T' &&
                transmissionData[1] == 'X' &&
                transmissionData[2] == 'L' &&
                transmissionData[3] == 'E' &&
                transmissionData[4] == 'D'
            )
            {
                return TRANSMISSION_IN_LEDMANAGER;
            }
            if (
                transmissionData[0] == 'T' &&
                transmissionData[1] == 'X' &&
                transmissionData[2] == 'P' &&
                transmissionData[3] == 'A' &&
                transmissionData[4] == 'N'
            )
            {
                return TRANSMISSION_IN_PANEL;
            }
            if (
                transmissionData[0] == 'T' &&
                transmissionData[1] == 'X' &&
                transmissionData[2] == 'D' &&
                transmissionData[3] == 'I' &&
                transmissionData[4] == 'O'
            )
            {
                return TRANSMISSION_IN_DIODE;
            }
            if (
                transmissionData[0] == 'I' &&
                transmissionData[1] == 'D' &&
                transmissionData[2] == 'E' &&
                transmissionData[3] == 'N' &&
                transmissionData[4] == 'T'
            )
            {
                return TRANSMISSION_IN_IDENT;
            }

            return TRANSMISSION_IN_NONE;
        }
        private byte waitAndRead()
        {
            while (serial.BytesToRead == 0)
            {


                Thread.Sleep(1);
                timeoutTimer++;
                if (timeoutTimer == TIMEOUTATTEMPTS)
                {
                    Console.WriteLine("\n[WaitAndRead timed out]");
                    timeoutTimer = 0;
                    return (0);
                }
            }

            for (int i = 0; i < (BUFFERSIZE - 1); i++)
            {
                transmissionData[i] = transmissionData[i + 1];
            }
            transmissionData[BUFFERSIZE - 1] = (byte)serial.ReadByte();
            Console.Write((char)transmissionData[BUFFERSIZE - 1]);


            return transmissionData[BUFFERSIZE - 1];
        }
        private bool doTransmissionEndCheck()
        {
            for (byte i = 0; i < BUFFERSIZE; i++)
            {
                waitAndRead();
                if
                (
                    transmissionData[0] == 'C' &&
                    transmissionData[1] == 'l' &&
                    transmissionData[2] == 'e' &&
                    transmissionData[3] == 'a' &&
                    transmissionData[4] == 'r'
                )
                {
                    Console.WriteLine("[LINK-CLR-ACK]");
                    return true;
                }
            }

            printBuffer();

            
            Console.WriteLine("[LINK-CLR-NCK]");
            return false;
        }
        private void printBuffer()
        {
            Console.Write("[LINK-BUF-PRI: ");
            for (byte i = 0; i < BUFFERSIZE; i++)
            {
                Console.Write((char)transmissionData[i]);
            }
            Console.WriteLine("]");
        }
        private void disconnect()
        {
            Console.WriteLine("Closed port " + serial.PortName);
            connected = false;
            serial.Close();
        }
        //Public
        public void manuallyConnect(int baudrate, string port)
        {
            if (serial.IsOpen)
            {
                disconnect();
            }
            serial.PortName = port;
            serial.BaudRate = baudrate;
            serial.Open();
            connected = true;
            Console.WriteLine("Manually connected to " + port);
        }
        public void autoConnect(int baudrate)
        {
            ports = SerialPort.GetPortNames();
            serial.BaudRate = baudrate;
            currentConnectingPort = -1; //-1 Because a ++ will set it to 0
        }
        public byte autoConnect_tick() //0:Failed, 1:Success, 2:Busy.
        {

            if (currentlyConnecting)
            {
                if (tick() == false)
                {
                    Console.WriteLine("AUTOCONNECT ROUND NEGATIVE");
                    serial.Close();
                    currentlyConnecting = false;
                }
                if (getReadyTransmissionType() == TRANSMISSION_IN_IDENT)
                {
                    connected = true;
                    Console.WriteLine("AUTOCONNECT SUCCESSFUL!!!");
                    return AUTOCONNECT_SUCCESS;
                }
            }
            else
            {
                currentConnectingPort++;
                if (currentConnectingPort == ports.Length)
                {
                    Console.WriteLine("AUTOCONNECT FAILED");
                    return AUTOCONNECT_FAILED;
                }
                else
                {
                    disconnect();
                    Console.WriteLine("AUTOCONNECT ATTEMPTING " + ports[currentConnectingPort]);
                    serial.PortName = ports[currentConnectingPort];
                    serial.Open();
                    transmit(TRANSMISSION_OUT_IDENT, "");
                    currentlyConnecting = true;
                    waitAndRead();
                }
            }
            return AUTOCONNECT_BUSY;
        }
        public bool tick() //0:Nothing, 1:All data was probably read.
        {
            if (serial.IsOpen || connected)
            {
                try
                {
                    while (serial.BytesToRead != 0)
                    {
                        for (int i = 0; i < (BUFFERSIZE - 1); i++)
                        {
                            transmissionData[i] = transmissionData[i + 1];
                        }
                        transmissionData[BUFFERSIZE - 1] = (byte)serial.ReadByte();
                        Console.Write((char)transmissionData[BUFFERSIZE - 1]);

                        //printBuffer();

                        byte receivedTransmissionType = decodeTransmissionType();

                        /*
                        if (receivedTransmissionType != TRANSMISSION_IN_NONE)
                        {
                            Console.Write("Transmission ");
                            Console.Write(receivedTransmissionType);
                        }
                        */

                        switch (receivedTransmissionType)
                        {
                            case TRANSMISSION_IN_LEDMANAGER:
                                Console.Write("[LINK-DET-MAN]");

                                buffer_ledManager.brightness = waitAndRead();
                                buffer_ledManager.speed = waitAndRead();
                                buffer_ledManager.enabled = Convert.ToBoolean(waitAndRead());
                                break;
                            case TRANSMISSION_IN_PANEL:
                                Console.Write("[LINK-DET-PAN]");

                                buffer_panel.number = waitAndRead();
                                buffer_panel.compassDir = waitAndRead();
                                buffer_panel.clockDir = Convert.ToBoolean(waitAndRead());
                                buffer_panel.diodeAmount = waitAndRead();

                                buffer_panel.brightness = waitAndRead();
                                buffer_panel.effect = waitAndRead();
                                buffer_panel.colour = waitAndRead();
                                buffer_panel.offset = waitAndRead();
                                buffer_panel.speed = waitAndRead();
                                buffer_panel.repeat = Convert.ToBoolean(waitAndRead());
                                buffer_panel.detailed = Convert.ToBoolean(waitAndRead());

                                buffer_panel.r = waitAndRead();
                                buffer_panel.g = waitAndRead();
                                buffer_panel.b = waitAndRead();
                                break;
                            case TRANSMISSION_IN_DIODE:
                                Console.Write("[LINK-DET-DIO]");
                                buffer_panel.number = waitAndRead();

                                buffer_panel.brightness = waitAndRead();
                                buffer_panel.effect = waitAndRead();
                                buffer_panel.colour = waitAndRead();
                                buffer_panel.offset = waitAndRead();
                                buffer_panel.speed = waitAndRead();
                                buffer_panel.repeat = Convert.ToBoolean(waitAndRead());

                                buffer_panel.r = waitAndRead();
                                buffer_panel.g = waitAndRead();
                                buffer_panel.b = waitAndRead();
                                break;
                            case TRANSMISSION_IN_IDENT:
                                Console.Write("[LINK-DET-ID]");
                                break;
                        }

                        if (receivedTransmissionType != TRANSMISSION_IN_NONE)
                        {
                            if (doTransmissionEndCheck())
                            {
                                readyTransmissionType = receivedTransmissionType;
                                Console.WriteLine("[LINK-TRANS-ACK]");
                            }
                            else
                            {
                                Console.WriteLine("[LINK-TRANS-ACK]");
                            }
                        }

                        while (serial.BytesToRead == 0)
                        {
                            Thread.Sleep(1);
                            timeoutTimer++;
                            if (timeoutTimer == TIMEOUTATTEMPTS)
                            {
                                Console.WriteLine("\n[Tick serial timed out]");
                                timeoutTimer = 0;
                                return true;
                            }
                        }
                    }
                }
                catch (System.InvalidOperationException)
                {
                    disconnect();
                }
            }
            return false;
        }
        public void transmit(byte transmissionType, string data)
        {
            switch (transmissionType)
            {
                case TRANSMISSION_OUT_PANELFX:
                    serial.Write("PanFx");
                    break;
                case TRANSMISSION_OUT_PANELCUSTOM:
                    serial.Write("PanCu");
                    break;
                case TRANSMISSION_OUT_DIODEFX:
                    serial.Write("DioFx");
                    break;
                case TRANSMISSION_OUT_DIODECUSTOM:
                    serial.Write("DioCu");
                    break;
                case TRANSMISSION_OUT_BRIGHTNESS:
                    serial.Write("Brigh");
                    break;
                case TRANSMISSION_OUT_SLEEPTIMER:
                    serial.Write("Sleep");
                    break;
                case TRANSMISSION_OUT_LIGHTSENSOR:
                    serial.Write("Light");
                    break;
                case TRANSMISSION_OUT_REQUEST:
                    serial.Write("Reque");
                    break;
                case TRANSMISSION_OUT_IDENT:
                    serial.Write("Ident");
                    break;
            }
            serial.Write(data);
            serial.Write("Clear");
        }
        public byte getReadyTransmissionType()
        {
            if (readyTransmissionType == TRANSMISSION_IN_IDENT)
            {
                readyTransmissionType = TRANSMISSION_IN_NONE;
                return TRANSMISSION_IN_IDENT;
            }
            return readyTransmissionType;
        }
        public string getConnectedPort()
        {
            return serial.PortName;
        }
        public bool getConnected()
        {
            return connected;
        }
        public string[] getPorts()
        {
            ports = SerialPort.GetPortNames();
            return ports;
        }
        public Transmission_LedManager getTransmission_LedManager()
        {
            return buffer_ledManager;
        }
        public Transmission_Panel getTransmission_Panel()
        {
            return buffer_panel;
        }
        public Transmission_Diode getTransmission_Diode()
        {
            return buffer_diode;
        }
        

        //Old Code
        public bool getConnectionClearance() //Checks if the system is connected properly in order to safely transmit.
        {
            throw new NotImplementedException();
        }
        public void serialWrite(string text) //Prints to the serial monitor
        {
            throw new NotImplementedException();
        }

    }
}

File: Codebase/Trinity Core MK3/src/Trinity/ControlPanel/Comms/Comms.cpp
#include "Comms.h"

Comms::Comms()
{
    Serial.println(F("Comms: Starting"));

    readyTransmissionType = TRANSMISSION_IN_NONE;

    buffer_LedManager       = new struct Transmission_LedManager;
    buffer_Panel            = new struct Transmission_Panel;
    buffer_Diode            = new struct Transmission_Diode;
    buffer_CustomPalette    = new struct Transmission_CustomPalette;
    for (uint8_t i = 0; i < AMOUNTOFCOLOURS; i++) buffer_CustomPalette->customRGB[i]= new struct ColourRGB;
    buffer_SleepTimer       = new struct Transmission_SleepTimer;
    buffer_LightSensorOffset= 0;

    Serial.println(F("Comms: Started"));
}

//Private
uint8_t Comms::decodeTransmissionType()
{
    if(!memcmp(transmissionData, "Reque", IDENTLENGTH)) return TRANSMISSION_IN_REQUEST;
    if(!memcmp(transmissionData, "Ident", IDENTLENGTH)) return TRANSMISSION_IN_IDENT;
    if(!memcmp(transmissionData, "LedMa", IDENTLENGTH)) return TRANSMISSION_IN_LEDMANAGER;
    if(!memcmp(transmissionData, "Panel", IDENTLENGTH)) return TRANSMISSION_IN_PANEL;
    if(!memcmp(transmissionData, "Diode", IDENTLENGTH)) return TRANSMISSION_IN_DIODE;
    if(!memcmp(transmissionData, "Custo", IDENTLENGTH)) return TRANSMISSION_IN_CUSTOMPALETTES;
    if(!memcmp(transmissionData, "Sleep", IDENTLENGTH)) return TRANSMISSION_IN_SLEEPTIMER;
    if(!memcmp(transmissionData, "Light", IDENTLENGTH)) return TRANSMISSION_IN_LIGHTSENSOR;
    return TRANSMISSION_IN_NONE;
}
uint8_t Comms::waitAndRead()
{
    while (!Serial.available()) delay(1);
    return (Serial.read());
}
bool    Comms::doTransmissionEndCheck()
{
    for(uint8_t i = 0; i < IDENTLENGTH; i++)
    {
        transmissionData[i] = waitAndRead();
    }
    
    //printBuffer();
    if(!memcmp(transmissionData, "Clear", IDENTLENGTH)) return true;

    return false;
}
void    Comms::printBuffer()
{
    Serial.println(F("[TRI-PRINTBUF "));
    for(uint8_t i = 0; i < IDENTLENGTH; i++)
    {
        Serial.write(transmissionData[i]);
    }
    Serial.println("]");
}

//Public
void                            Comms::tick()
{
    while(Serial.available() > 0 && readyTransmissionType == TRANSMISSION_IN_NONE)
    {
        for(byte i = 0; i < IDENTLENGTH-1; i++)
        {
            transmissionData[i] = transmissionData[i+1];
        }
        transmissionData[IDENTLENGTH-1] = Serial.read();

        uint8_t receivedTransmissionType = decodeTransmissionType();
        //Serial.print(F("Transmission "));
        //Serial.println(receivedTransmissionType);

        switch (receivedTransmissionType)
        {
            case TRANSMISSION_IN_LEDMANAGER:
            {
                buffer_LedManager->brightness        = waitAndRead();
                buffer_LedManager->speed            = waitAndRead();
            }
            break;
            case TRANSMISSION_IN_PANEL:
            {
                buffer_Panel->panelNumber         = waitAndRead();
                buffer_Panel->vfxData.effect      = waitAndRead();
                buffer_Panel->vfxData.colour      = waitAndRead();
                buffer_Panel->vfxData.offset      = waitAndRead();
                buffer_Panel->vfxData.speed       = waitAndRead();
                buffer_Panel->vfxData.repeat      = true;
                Serial.println("DD");
            }
            break;
            
            case TRANSMISSION_IN_DIODE:
            {
                buffer_Diode->panelNumber         = waitAndRead();
                buffer_Diode->diodeNumber         = waitAndRead();
                buffer_Diode->vfxData.effect      = waitAndRead();
                buffer_Diode->vfxData.colour      = waitAndRead();
                buffer_Diode->vfxData.offset      = waitAndRead();
                buffer_Diode->vfxData.speed       = waitAndRead();
                buffer_Diode->vfxData.repeat      = true;
                Serial.println("DD");
            }
            break;
            case TRANSMISSION_IN_CUSTOMPALETTES:
            {
                buffer_CustomPalette->slot = waitAndRead();

                for (uint8_t i = 0; i < AMOUNTOFCOLOURS; i++)
                {
                    buffer_CustomPalette->customRGB[i]->r = waitAndRead();
                    buffer_CustomPalette->customRGB[i]->g = waitAndRead();
                    buffer_CustomPalette->customRGB[i]->b = waitAndRead();
                }
            }
            break;
            case TRANSMISSION_IN_SLEEPTIMER:
            {
                buffer_SleepTimer->timerID   = waitAndRead();
                buffer_SleepTimer->hour      = waitAndRead();
                buffer_SleepTimer->minute    = waitAndRead();
                buffer_SleepTimer->enabled   = waitAndRead();
            }
            break;
            case TRANSMISSION_IN_LIGHTSENSOR:
            {
                buffer_LightSensorOffset = waitAndRead();
            }
            break;
        }

        if(receivedTransmissionType != TRANSMISSION_IN_NONE)
        {
            if (doTransmissionEndCheck())
            {
                readyTransmissionType = receivedTransmissionType;
                Serial.println(F("[TRI-TRANS-ACK]"));
            }
            else
            {
                Serial.println(F("[TRI-TRANS-ERR]"));
            }
        }
        

        if(!Serial.available())
        {            
            delay(1); //Wait a bit in case the transmisison was not actually finished
        }
    }
}
void                            Comms::transmit(String data)
{
    Serial.print(data);
    Serial.println("Clear");
}
uint8_t                         Comms::getReadyTransmissionType()
{
    if(readyTransmissionType == TRANSMISSION_IN_REQUEST || readyTransmissionType == TRANSMISSION_IN_IDENT)
    {
        uint8_t transmissionTypeToSend = readyTransmissionType;
        readyTransmissionType = TRANSMISSION_IN_NONE;
        return transmissionTypeToSend;
    }
    return readyTransmissionType;
}
Transmission_LedManager         Comms::getTransmission_LedManager()
{
    readyTransmissionType = TRANSMISSION_IN_NONE;
    return *buffer_LedManager;
}
Transmission_Panel              Comms::getTransmission_Panel()
{
    readyTransmissionType = TRANSMISSION_IN_NONE;
    return *buffer_Panel;
}
Transmission_Diode              Comms::getTransmission_Diode()
{
    readyTransmissionType = TRANSMISSION_IN_NONE;
    return *buffer_Diode;
}
Transmission_CustomPalette      Comms::getTransmission_CustomPalette()
{
    readyTransmissionType = TRANSMISSION_IN_NONE;
    return *buffer_CustomPalette;
}
Transmission_SleepTimer         Comms::getTransmission_SleepTimer()
{
    readyTransmissionType = TRANSMISSION_IN_NONE;
    return *buffer_SleepTimer;
}
int8_t                          Comms::getTransmission_LightSensorOffset()
{
    readyTransmissionType = TRANSMISSION_IN_NONE;
    return buffer_LightSensorOffset;
}